pkg/semtech-loramac: model in Kconfig

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

As the title says.

To fully match the Make dependency resolution, I had to modify the sx127x Kconfig file by removing the usage of choices. Maybe this is wrong but I don't see how to properly include `MODULE_SX1276` to the build with Kconfig. Looking at `tests/driver_sx127x` it seems to be possible by pulling in both `MODULE_SX1276` and `MODULE_SX127X`. That seems bogus and counter intuitive to me as I would expect the sx127x moduie to be pulled in automatically like it's done by this PR.

Note that this PR won't pass Murdock because of a mismatch resolution of the random module dependencies (again a choice is in the loop).

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

Green Murdock

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

#16875

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

leandrolanzieri commented on drivers/sx127x/Kconfig:
```diff
@@ -25,33 +25,22 @@ menuconfig MODULE_SX127X
-
 config MODULE_SX1272
     bool "SX1272"
+    select MODULE_SX127X
```
The usage of choices for the variant was a design decision for all drivers. We know that it requires to select both the driver module and variant, but that is only needed when the variant is to be forced. In case the board has the device connected to it, the `HAVE_*` symbol would be selected, indicating a preference for the choice. 

We have considered the approach you have here in the past (actually we started modelling like that), but we realized that it would not allow us to use modules that enable defaults (`MODULE_NETDEV_DEFAULT` in this case). Now, when such default is enabled, the driver module would be enabled (if the correct feature is present) and the correct variant would be the default choice option. This can't be done when the entry-point is the choice. There is some info in this PR https://github.com/RIOT-OS/RIOT/pull/17556/files#diff-2132175f34702f4d76287f4df5c5f5db02d097c2778e5ac47da4c88ef8f13acdR517.

aabadie replied:
From the board with hardwired radio perspective that could make sense. From the end user that plugged the radio on its board, I think it doesn't make sense since he won't have the `HAVE_` symbol defined for his board (thinking of a nucleo64 where one plugs an mbed lora shield), resulting in the user having to enable both sx127x and sx1276 (or sx1272 depending on the shield he is using).

Offline @fjmolinas told me that the choice was here to prevent both sx1272 and sx1276 to be enabled at the same time. But in make one could do that already. I would rather stick to the same behavior as make as a first step and after think of solutions for this problem later.

aabadie commented on drivers/sx127x/Kconfig:
```diff
@@ -25,33 +25,22 @@ menuconfig MODULE_SX127X
 config HAVE_SX1272
     bool
-    select HAVE_SX127X
+    select MODULE_SX1272
```
Thinking of this again, this should rather be:
```suggestion
    select MODULE_SX1272 if MODULE_NETDEV_DEFAULT
```

MrKevinWeiss reviewed (changes requested):
It would be good if we keep the same design here.

I know that it is a bit verbose to require both the module and the variant explicitly since the variant would require the module.

This has been played with for quite some time now and it the best solution given the constraints we are dealing with.

I strongly push the `HAVE_*` because I think having a verbose (and machine readable) description of the board is a good thing.

We can open up the conversation and look for a better way.

aabadie commented:
> I strongly push the `HAVE_*` because I think having a verbose (and machine readable) description of the board is a good thing.

I have no objection with the `HAVE_` and I even find this concept quite nice. If you see my latest change, `MODULE_SX1276` symbol can be pull-in automatically by `HAVE_SX1276` if `MODULE_NETDEV_DEFAULT` is selected.
This is what looks the most natural to me and this allows the automatic inclusion of `MODULE_SX127X` from the specific driver module.

aabadie commented:
I removed the Kconfig changes related to sx127x driver and added the missing symbols in loramac applications config. Locally the dependency resolution is the same as with Make so I have good hopes that it will pass Murdock now.

MrKevinWeiss commented on examples/lorawan/Kconfig:
```diff
@@ -0,0 +1,5 @@
+    bool
+    default y
+    depends on TEST_KCONFIG
+    select MODULE_PERIPH_RTC if HAS_PERIPH_RTC
```
Maybe this could help

```suggestion
    select MODULE_PERIPH_RTC if HAS_PERIPH_RTC && !HAVE_SHARED_PERIPH_RTT_PERIPH_RTC
```

MrKevinWeiss reviewed (commented):
I don't know if you want to note this somewhere but after the migration is complete we can probably remove most of the `CONFIG_MODULE_SX1276=y` and just default be that but if boards have something different then it should use it.
This is only be we need to match the make system and currently it is not implemented.

MrKevinWeiss commented:
Heh, me too.  I will take another look at it today.

MrKevinWeiss commented:
> I'm a bit lost here, I need some refresh. @MrKevinWeiss what's the situation with the random module ?

So far I don't see issues with the random module.  The real issue is with the ztimer.  I will probably try to open something else because it seems the core of the problem is with the ztimer backend selection.
- We have 3 backends (`periph_rtt`, `periph_rtc`, and `periph_timer`)
- Some boards have a feature conflict when using `periph_rtt` and `periph_rtc` indicated by `HAVE_SHARED_PERIPH_RTT_PERIPH_RTC`
- This application would want to use `periph_rtc` if possible as it allows extra features

`make` seems to select using `periph_timer` for msec and `periph_rtc` for sec, optimizing for features, while `Kconfig` seems to prefer the `periph_rtt`. To throw a wrench in the cog, native has some strange behaviour with circular dependencies of ztimer and periph_rtc as it simulates periph_rtc with ztimer that can use periph_rtc as a backend...

There are quite some comments about the this in make, also referring to kconfig.

Unfortunately it will take some time to figure out how to get matching behaviour and even more time to figure out ideal behaviour.

<details><summary>This fixed a few boards I was testing but I don't think this is a correct solution, if you want see what the CI says</summary>

```diff
diff --git a/examples/lorawan/Kconfig b/examples/lorawan/Kconfig
index 15b7b2c54c..5369220c31 100644
--- a/examples/lorawan/Kconfig
+++ b/examples/lorawan/Kconfig
@@ -2,4 +2,12 @@ config APPLICATION
     bool
     default y
     depends on TEST_KCONFIG
-    select MODULE_PERIPH_RTC if HAS_PERIPH_RTC && (!HAVE_SHARED_PERIPH_RTT_PERIPH_RTC || !MODULE_PERIPH_RTT)
+    imply MODULE_PERIPH_RTC
+
+choice ZTIMER_SEC_BACKEND
+    default ZTIMER_SEC_BACKEND_RTC
+endchoice
+
+choice ZTIMER_MSEC_BACKEND
+    default ZTIMER_MSEC_BACKEND_TIMER
+endchoice
diff --git a/sys/ztimer/Kconfig b/sys/ztimer/Kconfig
index f1ab66248e..c7bbe2791b 100644
--- a/sys/ztimer/Kconfig
+++ b/sys/ztimer/Kconfig
@@ -59,7 +59,7 @@ config MODULE_ZTIMER_MSEC
     bool "Milliseconds"
     select MODULE_ZTIMER
 
-choice
+choice ZTIMER_MSEC_BACKEND
     bool "Backend"
     depends on MODULE_ZTIMER_MSEC
     default ZTIMER_MSEC_BACKEND_RTT if !MODULE_ZTIMER_NO_PERIPH_RTT
@@ -79,7 +79,7 @@ config MODULE_ZTIMER_SEC
     bool "Seconds"
     select MODULE_ZTIMER
 
-choice
+choice ZTIMER_SEC_BACKEND
     bool "Backend"
     depends on MODULE_ZTIMER_SEC
     default ZTIMER_SEC_BACKEND_RTT if !MODULE_ZTIMER_NO_PERIPH_RTT
```
</details>

MrKevinWeiss commented:
I haven't forgotten about this but there really is some issues with Kconfig matching the make resolution of the backends for ztimer. 
Just having the following will bring the issue to light...
```
USEMODULE += ztimer
USEMODULE += ztimer_msec
USEMODULE += ztimer_sec

FEATURES_OPTIONAL += periph_rtc
```
I wonder if for now we just hardcode the backends to something that works?

aabadie commented:
I need help to fix the kconfig circular dependency here. @MrKevinWeiss any idea ?

MrKevinWeiss commented:
So first I will say the issue is not with the semtech package but with the ztimer and underlying peripheral modeling.

simply having a
```makefile
FEATURES_OPTIONAL += periph_rtc
USEMODULE += ztimer_msec
```

running against

```kconfig
config APPLICATION
    bool
    default y
    imply MODULE_PERIPH_RTC if TEST_KCONFIG
```

```
CONFIG_MODULE_ZTIMER_MSEC=y
CONFIG_MODULE_ZTIMER=y
```

will result in differences.

These differences are mainly boards selecting the `ztimer_periph_timer` backend in `make` and selecting the `ztimer_periph_rtt` backend in kconfig.  This is due to kconfig not handling the defaulting correctly when the `HAVE_SHARED_PERIPH_RTT_PERIPH_RTC` since we would prefer to use `rtc` features requiring using the `ztimer_periph_timer` rather then no features but the `ztimer_periph_rtt`.

I have seen other instances where `make` will select the `periph_timer` backend for seemingly no reason...  I would assume if everything is OK that we would prefer to use rtt over the timer.

The whole system needs a major rework, the "no circular dependencies` makes it really hard, especially when dealing with how native works with the rtc.

As yet-another-workaround we can try the following diff which simply forces the periph_timer as the backend in all cases.

```diff
diff --git a/boards/xg23-pk6068a/Makefile.dep b/boards/xg23-pk6068a/Makefile.dep
index 01d21e3373..abc3287a7b 100644
--- a/boards/xg23-pk6068a/Makefile.dep
+++ b/boards/xg23-pk6068a/Makefile.dep
@@ -3,5 +3,7 @@ ifneq (,$(filter saul_default,$(USEMODULE)))
 endif
 
 ifneq (,$(filter ztimer_msec ztimer_sec,$(USEMODULE)))
-  USEMODULE += ztimer_periph_lptimer
+  ifeq (,$(filter ztimer_periph_timer,$(USEMODULE)))
+    USEMODULE += ztimer_periph_lptimer
+  endif
 endif
diff --git a/examples/lorawan/Kconfig b/examples/lorawan/Kconfig
index 15b7b2c54c..8a6ea6f496 100644
--- a/examples/lorawan/Kconfig
+++ b/examples/lorawan/Kconfig
@@ -1,5 +1,8 @@
 config APPLICATION
     bool
     default y
-    depends on TEST_KCONFIG
-    select MODULE_PERIPH_RTC if HAS_PERIPH_RTC && (!HAVE_SHARED_PERIPH_RTT_PERIPH_RTC || !MODULE_PERIPH_RTT)
+    imply MODULE_PERIPH_RTC if TEST_KCONFIG
+
+choice ZTIMER_MSEC_BACKEND
+    default ZTIMER_MSEC_BACKEND_TIMER
+endchoice
diff --git a/examples/lorawan/Makefile b/examples/lorawan/Makefile
index 67a1243c81..c84723617f 100644
--- a/examples/lorawan/Makefile
+++ b/examples/lorawan/Makefile
@@ -45,6 +45,9 @@ USEMODULE += auto_init_loramac
 USEMODULE += $(DRIVER)
 USEMODULE += fmt
 FEATURES_OPTIONAL += periph_rtc
+# Needed to fix the backend o ztimer so the RTC is free
+USEMODULE += ztimer_msec
+USEMODULE += ztimer_periph_timer
 
 # Uncomment the following line to enable Loramac stack state persistence on EEPROM.
 # Make sure the EEPROM is erased before enabling this and when flashing a board
diff --git a/examples/lorawan/app.config.test b/examples/lorawan/app.config.test
index 3d082b221c..9b31cb1e31 100644
--- a/examples/lorawan/app.config.test
+++ b/examples/lorawan/app.config.test
@@ -4,3 +4,7 @@ CONFIG_MODULE_SX127X=y
 CONFIG_MODULE_PRNG_MUSL_LCG=y
 CONFIG_MODULE_AUTO_INIT_LORAMAC=y
 CONFIG_MODULE_FMT=y
+
+CONFIG_MODULE_ZTIMER_MSEC=y
+CONFIG_MODULE_ZTIMER_PERIPH_TIMER=y
+CONFIG_MODULE_ZTIMER=y
diff --git a/sys/ztimer/Kconfig b/sys/ztimer/Kconfig
index b9b56d422d..38669dd5a1 100644
--- a/sys/ztimer/Kconfig
+++ b/sys/ztimer/Kconfig
@@ -71,6 +71,7 @@ config MODULE_ZTIMER_MSEC
 choice ZTIMER_MSEC_BACKEND
     bool "Backend"
     depends on MODULE_ZTIMER_MSEC
+    default ZTIMER_MSEC_BACKEND_TIMER if MODULE_ZTIMER_USEC
     default ZTIMER_MSEC_BACKEND_LPTIMER if HAVE_ZTIMER_PERIPH_LPTIMER
     default ZTIMER_MSEC_BACKEND_RTT if !MODULE_ZTIMER_NO_PERIPH_RTT
     default ZTIMER_MSEC_BACKEND_TIMER
diff --git a/sys/ztimer/Makefile.dep b/sys/ztimer/Makefile.dep
index eec51b70b0..ae7daa3f13 100644
--- a/sys/ztimer/Makefile.dep
+++ b/sys/ztimer/Makefile.dep
@@ -87,7 +87,7 @@ endif
 
 ifneq (,$(filter ztimer_msec ztimer_sec,$(USEMODULE)))
   USEMODULE += ztimer
-  ifeq (,$(filter ztimer_no_periph_rtt,$(USEMODULE)))
+  ifeq (,$(filter ztimer_no_periph_rtt ztimer_periph_timer,$(USEMODULE)))
     FEATURES_OPTIONAL += periph_rtt
     # HACK: periph_rtt will get used only in the next iteration but an updated
     # state for FEATURES_USED is needed here so include `features_check.inc.mk`
```

MrKevinWeiss commented on sys/ztimer/Kconfig:
```diff
@@ -71,6 +71,7 @@ config MODULE_ZTIMER_MSEC
 choice ZTIMER_MSEC_BACKEND
     bool "Backend"
     depends on MODULE_ZTIMER_MSEC
+    default ZTIMER_MSEC_BACKEND_TIMER if MODULE_ZTIMER_USEC
```
Ahh yes, native is fun to model.  Just create a different symbol, something like `PREFER_ZTIMER_PERIPH_TIMER` so it would look like:

```diff
diff --git a/sys/ztimer/Kconfig b/sys/ztimer/Kconfig
index 38669dd5a1..277a502191 100644
--- a/sys/ztimer/Kconfig
+++ b/sys/ztimer/Kconfig
@@ -60,6 +60,10 @@ config ZTIMER_USEC
     select MODULE_ZTIMER_USEC
     select MODULE_ZTIMER_PERIPH_TIMER
     select MODULE_ZTIMER
+    select PREFER_ZTIMER_PERIPH_TIMER
+
+config PREFER_ZTIMER_PERIPH_TIMER
+    bool
 
 config MODULE_ZTIMER_USEC
     bool
@@ -71,7 +75,7 @@ config MODULE_ZTIMER_MSEC
 choice ZTIMER_MSEC_BACKEND
     bool "Backend"
     depends on MODULE_ZTIMER_MSEC
-    default ZTIMER_MSEC_BACKEND_TIMER if MODULE_ZTIMER_USEC
+    default ZTIMER_MSEC_BACKEND_TIMER if PREFER_ZTIMER_PERIPH_TIMER
     default ZTIMER_MSEC_BACKEND_LPTIMER if HAVE_ZTIMER_PERIPH_LPTIMER
     default ZTIMER_MSEC_BACKEND_RTT if !MODULE_ZTIMER_NO_PERIPH_RTT
     default ZTIMER_MSEC_BACKEND_TIMER
@@ -99,6 +103,7 @@ config MODULE_ZTIMER_SEC
 choice
     bool "Backend"
     depends on MODULE_ZTIMER_SEC
+    default ZTIMER_SEC_BACKEND_TIMER if PREFER_ZTIMER_PERIPH_TIMER
     default ZTIMER_SEC_BACKEND_LPTIMER if HAVE_ZTIMER_PERIPH_LPTIMER
     default ZTIMER_SEC_BACKEND_RTT if !MODULE_ZTIMER_NO_PERIPH_RTT
     default ZTIMER_SEC_BACKEND_TIMER
```

MrKevinWeiss commented:
Ugg, I will look into it on Monday...

MrKevinWeiss commented:
Here is the TLDR fix:
```diff
diff --git a/tests/drivers/io1_xplained/Makefile b/tests/drivers/io1_xplained/Makefile
index 5f91154c3a..dcbb6c54ec 100644
--- a/tests/drivers/io1_xplained/Makefile
+++ b/tests/drivers/io1_xplained/Makefile
@@ -2,5 +2,7 @@ include ../Makefile.drivers_common
 
 USEMODULE += io1_xplained
 USEMODULE += ztimer_msec
+# Hack to prevent some boards from bring in rtt and periph_timer backend
+USEMODULE += ztimer_periph_timer
 
 include $(RIOTBASE)/Makefile.include
 ```

It is because the `MODULE_IO1_XPLAINED` brings in `MODULE_SDCARD_SPI` which needs `ZTIMER_USEC`, and I guess due to the order of inclusion in make the first round of the `ztimer_msec` brings in the rtt backend, then when it brings in the `ztimer_usec` the `periph_timer` gets used as the backend.  I am assuming we prefer just the one if possible so would just hardcode it to use the periph_timer as the backend and make a note of it.

Once all the xtimer is removed and if we can finish the kconfig migration then we can introduce some policy selection or something that would let us optimize for code size vs power savings.

aabadie commented:
> Here is the TLDR fix

Thanks, let's see [what Murdock has to say](https://ci.riot-os.org/details/5c0c20bf6b8744e9918aeaafbea747d0)

MrKevinWeiss commented:
Ahh no fail fast, good... if this really is the last one then we can go. If there are more than maybe there needs to be a better way.

MrKevinWeiss commented:
Hmm it seems an issue with bors merging multiple PRs... I wonder if we can just get it isolated...

MrKevinWeiss commented:
I guess it is because the first iteration of make has `ztimer_msec` and nothing else, which brings in the rtt.  The later iterations bring in xtimer which brings in usec which brings in periph_timer.  Since the first round cannot know periph timer will be used we are forced to bring in both.

I wonder what the ideal behaviour is, use both backeds (since the msec will be able to sleep) or stop with this, if I am using a backend for something else then just use that one (which saves space but prevents sleep).

aabadie commented:
Murdock is finally green here!

MrKevinWeiss reviewed (approved):
ACK, lets see what a larger test will say but I think we need to move on this one.  This does change make to have a more consistent ztimer backend selection...  It might have some effect for those who assumed power savings over code size.

aabadie commented:
bors cancel

(trying to fill a merge :train: )

aabadie commented:
bors merge

aabadie commented:
bors merge

aabadie commented:
bors cancel

aabadie commented:
bors merge

aabadie commented:
bors cancel

Some concerns about ztimer are rising in private.

MrKevinWeiss commented:
Maybe I will try different PR that will only deal with the problems with ztimer, it is out of scope of this PR and should probably be blocklisted or wait until the ztimer pr is done.

aabadie commented:
This PR is no longer needed
